### PR TITLE
[JENKINS-31574] Improved validation for Scan Credentials

### PR DESCRIPTION
**Pull request by recena**
[JENKINS-31574](https://issues.jenkins-ci.org/browse/JENKINS-31574)

- [x] Improve error handling
- [x] Anonymous (none) is a option valid for Scan Credentials
- [x] Implement a `RateLimitHandler`. Used in anonymous connections.

### Form validation

jenkins-31574_1

### Behavior with error handling in place

jenkins-31574_2

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 1]
```diff
@@ -107,8 +109,12 @@ public String getApiUri() {
             listener.getLogger().format("No scan credentials, skipping%n");
             return;
         }
-        listener.getLogger().format("Connecting to GitHub using %s%n", CredentialsNameProvider.name(credentials));
         GitHub github = Connector.connect(apiUri, credentials);
+        if (!github.isCredentialValid()) {
```
Comment (tfennelly) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: I assume `github` is `@Nonnull`?

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 2]
```diff
@@ -107,8 +109,12 @@ public String getApiUri() {
...
+            listener.getLogger().format("Invalid scan credentials, skipping%n");
+            return;
+        }
+        listener.getLogger().format("Connecting to GitHub using %s%n", CredentialsNameProvider.name(credentials));
```
Comment (tfennelly) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: Do you need to add the `%n` at the end? Just curious.

**Comment by tfennelly** [comment]
:bee: insofar as I get it :)

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 3]
```diff
@@ -107,8 +109,12 @@ public String getApiUri() {
             listener.getLogger().format("No scan credentials, skipping%n");
             return;
         }
-        listener.getLogger().format("Connecting to GitHub using %s%n", CredentialsNameProvider.name(credentials));
         GitHub github = Connector.connect(apiUri, credentials);
+        if (!github.isCredentialValid()) {
+            listener.getLogger().format("Invalid scan credentials, skipping%n");
+            return;
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: This implies that credentials `- none -` is not valid the _Scan Credentials_ dropdown list. Do we want that?

**Comment by amuniz** [comment]
I would _bee_ this PR because it solves the UX issue described in the JIRA issue. But :ant: on allowing to select _- none -_ as _Scan Credentials_ when it's not really allowed in the backend (it will always print `Invalid scan credentials, skipping`).

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java (hunk 1 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: @tfennelly Yes. [Ref](https://github.com/jenkinsci/github-branch-source-plugin/blob/master/src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java#L68).

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java (hunk 2 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: @tfennelly Yes. We are trying to follow a format for logging messages.

**Comment by recena** [comment]
@amuniz I agree. If _Scan Credentials_ is required (everything points to), we should remove the `-none-``option.

**Comment by jglick** [comment]
I think you need a default value for the field. Otherwise anyone configuring the form for the first time will see the pulldown set to the first matching credentials item, typically something inappropriate.

There is already a form validation warning in case _none_ is selected, and the log already says

> No scan credentials, skipping

so I do not see any problem with leaving the natural default.

(And we might want to allow _none_ to actually work, though I would still recommend form validation noting that this is only an option for public owners, and you may be subjected to strict rate limiting.)

**Comment by recena** [comment]
@jglick The usability issue is when the user think about `none` as to `anonymous`. Well, in this case this PR is ready to merge.

**Comment by recena** [comment]
@jglick This PR does not allow work without a valid scan credentials. This means that an anonymous access is not possible. IIRC, we agree with this so that to avoid the [rate limit](https://developer.github.com/v3/#rate-limiting) in nauthenticated requests.

**Comment by recena** [comment]
To be pending....

```
/* TODO call GitHubBuilder withRateLimitHandler to notify listener so we do not get stuck without messages in something like
java.lang.Thread.State: TIMED_WAITING (sleeping)
        at java.lang.Thread.sleep(Native Method)
        at org.kohsuke.github.RateLimitHandler$1.onError(RateLimitHandler.java:38)
        at org.kohsuke.github.Requester.handleApiError(Requester.java:486)
        at org.kohsuke.github.Requester._to(Requester.java:245)
        at org.kohsuke.github.Requester.to(Requester.java:191)
        at org.kohsuke.github.GitHub.getRepository(GitHub.java:320)
*/
```

**Comment by recena** [comment]
@jglick @amuniz In my opinion, this PR is ready to merge.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java [hunk 4]
```diff
@@ -108,4 +111,25 @@ private static CredentialsMatcher githubScanCredentialsMatcher() {
 
     private Connector() {}
 
+    /**
+     * Fail immediately and throw a customized exception.
+     */
+    public static final RateLimitHandler CUSTOMIZED = new RateLimitHandler() {
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: Should not be this used for all connections? (missed in [line 77](https://github.com/jenkinsci/github-branch-source-plugin/pull/11/files#diff-61948c4cd4e480f18c6e572cf88a055bR77)) Probably nobody reaches the rate limit being authenticated, but who knows...

**Comment by amuniz** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java (hunk 4 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: @amuniz By the default all connections use `RateLimitHandler.WAIT`([ref](https://github.com/kohsuke/github-api/blob/github-api-1.71/src/main/java/org/kohsuke/github/GitHubBuilder.java#L32)). As part of this PR we have implemented a new RateLimitHandler used until now only for anonymous connection.

**Comment by amuniz** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java (hunk 4 above)
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: I resolved rate limit issues long time ago, WAIT is not only one available rate limit handler. I don't understand why are you spending so much time around making connections (the third PR to this plugin if i'm not mistaken).

Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: @KostyaSha Indeed, there is another one `RateLimitHandler.FAIL` but throw a generic `IOException`. This implementation (`CUSTOMIZED`) is more flexible and I can do things like this:

```
Proposing build-pipeline-plugin
Connecting to GitHub using anonymous access
Looking up recena/build-pipeline-plugin

  Getting remote branches...

    Checking branch JENKINS-31746
      Jenkinsfile does not exist in this branch
    Does not meet criteria


    Checking branch cleanup
      Jenkinsfile does not exist in this branch
    Does not meet criteria


    Checking branch master
      Jenkinsfile does not exist in this branch
    Does not meet criteria


    Checking branch recovery
      Jenkinsfile does not exist in this branch
    Does not meet criteria


    Checking branch retry-param-and-security-fixes
      Jenkinsfile does not exist in this branch
    Does not meet criteria


    Checking branch revert-66-master
      Jenkinsfile does not exist in this branch
    Does not meet criteria

  0 branches were processed

Done examining recena/build-pipeline-plugin

Proposing cartodb
Connecting to GitHub using anonymous access
Looking up recena/cartodb

  Getting remote branches...

GitHub API rate limit exceeded

Aborted
Finished: ABORTED
```

In this PR I'm improving details related to UX.

Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: RateLimit is the same connection problem as IOException, connecting to instances with anonymous access :-1:

Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: @KostyaSha I did not understand your comment. Could you help me to understand it?

Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: @KostyaSha do you mean that hitting RateLimit will result in an IOException?

Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: @abayer yes, it what exactly i fixed in github-api. Either it Thread.sleeps() with wait either fails. In latest github-api version something was changed in fail logic (order of checks).  I.e. if you provided wrong creds, you should be able to fail and pick new connector. (Most of issues was verified in my pr plugin and resolved in api).

Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: @abayer It should throw an `IOException`, in my case `RateLimitExceededException` extends of `IOException`.

Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: @KostyaSha This implementation works fine. Please, find in my comment a console output. I cannot see the problem.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    private static final long serialVersionUID = 1L;
+
+    private long limit;
+
+    private long remaining;
+
+    private long reset;
+
+    public RateLimitExceededException() {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: Why does this constructor exist?

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        super();
+    }
+
+    public RateLimitExceededException(String msg, long limit, long remaining, long reset) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: The `msg` need not be a constructor parameter. There is nothing for a caller to say other than, well, the rate limit was exceeded.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        this.limit = limit;
+        this.remaining = remaining;
+        this.reset = reset;
+    }
+
+    public RateLimitExceededException(Throwable cause) {
+        initCause(cause);
+    }
+
+    public RateLimitExceededException(String message, Throwable cause) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: :ant: Delete if unused.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    public RateLimitExceededException(String msg, long limit, long remaining, long reset) {
+        super(msg);
+        this.limit = limit;
+        this.remaining = remaining;
+        this.reset = reset;
+    }
+
+    public RateLimitExceededException(Throwable cause) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: :ant: Delete if unused.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        initCause(cause);
+    }
+
+    public RateLimitExceededException(String message, Throwable cause) {
+        super(message);
+        initCause(cause);
+    }
+
+    public long getReset() {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: :ant: all these getters seem unused. The `super` call should rather pass a customized `message` that includes relevant numbers.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java [hunk 10]
```diff
@@ -108,4 +111,25 @@ private static CredentialsMatcher githubScanCredentialsMatcher() {
...
+        public void onError(IOException e, HttpURLConnection uc) throws IOException {
+            try {
+                long limit = Long.parseLong(uc.getHeaderField("X-RateLimit-Limit"));
+                long remaining = Long.parseLong(uc.getHeaderField("X-RateLimit-Remaining"));
+                long reset = Long.parseLong(uc.getHeaderField("X-RateLimit-Reset"));
+
+                throw new RateLimitExceededException("GitHub API rate limit exceeded", limit, remaining, reset);
+            } catch (NumberFormatException nfe) {
+                // Something wrong happened
+                throw new IOException();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: :ant:

```java
throw new IOException(nfe);
```

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java [hunk 11]
```diff
@@ -108,4 +111,25 @@ private static CredentialsMatcher githubScanCredentialsMatcher() {
 
     private Connector() {}
 
+    /**
+     * Fail immediately and throw a customized exception.
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: Seems nice but should at some point be in `github-api`, so file an RFE for that plugin so we can eventually move this code upstream.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 12]
```diff
@@ -103,46 +105,63 @@ public String getApiUri() {
...
+
         GHUser user = null;
         try {
             user = github.getUser(repoOwner);
+        } catch (RateLimitExceededException rle) {
+            listener.getLogger().format("%n%s%n%n", rle.getMessage());
+            throw new InterruptedException();
         } catch (IOException e) {
-            // may be organization... ok to ignore
+            // Something wrong happened, maybe java.net.ConnectException?
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: :ant: Not part of this PR per se, but anyway: should log such things at least at `FINE`, so we can do diagnosis of problems in the field with a custom logger.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 13]
```diff
@@ -182,11 +201,26 @@ private void add(TaskListener listener, SCMSourceObserver observer, GHRepository
...
+                            return FormValidation.ok();
+                        }
+                    } catch (IOException e) {
+                        // ignore, never thrown
+                    }
+                }
+                return FormValidation.error("Invalid credentials");
             } else {
-                return FormValidation.warning("Credentials are required");
+                return FormValidation.warning("Credentials are recommended");
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: :ant: Should clarify to the user that they are recommended even for public repositories/organizations to avoid low rate limits.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 14]
```diff
@@ -182,11 +201,26 @@ private void add(TaskListener listener, SCMSourceObserver observer, GHRepository
...
+                if (credentials == null) {
+                    FormValidation.error("Invalid credentials");
+                } else {
+                    try {
+                        GitHub connector = Connector.connect(apiUri, credentials);
+                        if (connector.isCredentialValid()) {
+                            return FormValidation.ok();
+                        }
+                    } catch (IOException e) {
+                        // ignore, never thrown
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: Truly never thrown? Then wrap in `IllegalStateException`, or

```java
assert false : e;
```

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java (hunk 12 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: I will include a `logger` in a different PR. The change is related.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java (hunk 13 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: Well, this is only a validation message. That recommendation should be in the documentation.

**Comment by jglick** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java (hunk 13 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: Anyway, when the rate limit is exceeded the message is clear in the UI.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java (hunk 6 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: I prefer to leave the message open. The developer decides.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java (hunk 7 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: Maybe used in the future!

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java (hunk 8 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: Ditto

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java (hunk 9 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: What happens if I want to print the rest time to the use?

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java (hunk 10 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: Done.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java (hunk 11 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: I thought about it but I need to advance here.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java (hunk 14 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: I'll include a `LOGGER` in a separated PR.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java (hunk 9 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: Reset time I guess you mean. That was my suggestion, that the `RateLimitExceededException` constructor should accept the raw data, and pass an appropriately formatted message to the `super` constructor.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java (hunk 7 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: Never introduce APIs without a known use case. Adding them later if you need them is easy. Removing them later if you do not, or it turns out your design was flawed, is very hard.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java (hunk 9 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: Sure, *rest* is probably what I need.

I'm confused. That constructor is:
`public RateLimitExceededException(String msg, long limit, long remaining, long reset)`

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java (hunk 7 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: Well, I know these use cases but I have not yet implemented.

Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: Then you do not really know that your design was right.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java (hunk 9 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: My suggestion was that the constructor should take `(long, long, long)` and construct an appropriate `message` based on that information.

Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: As I said, I prefer the developer uses his own message.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java (hunk 7 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/RateLimitExceededException.java]: Really? You'll able to see it in next PR.
